ironmon_tracker/CustomCode.lua

```diff
@@ -0,0 +1,128 @@
+	CustomCode.objects = {}
+	local customCodeFolder = FileManager.getCustomFolderPath()
+	local filesLoaded = {
+		successful = {},
+		failed = {},
+	}
+
+	for _, filename in ipairs(CustomCode.filenames) do
+		local filepath = FileManager.getPathIfExists(customCodeFolder .. filename)
+		if filepath ~= nil then
+			local customObject = dofile(filepath)
+			if type(customObject) == "function" then
+				table.insert(CustomCode.objects, customObject() or {})
+			elseif type(customObject) == "table" then
+				table.insert(CustomCode.objects, customObject)
+			end
+			table.insert(filesLoaded.successful, filename)
+		else
+			table.insert(filesLoaded.failed, filename)
+		end
+	end
+
+	if #filesLoaded.successful > 0 then
+		print(string.format("%s: %s", CustomCode.Labels.filesLoadSuccess, table.concat(filesLoaded.successful, ", ")))
+	end
+	if #filesLoaded.failed > 0 then
+		print(string.format("%s: %s", CustomCode.Labels.filesLoadFailure, table.concat(filesLoaded.failed, ", ")))
+	end
+
+	CustomCode.execFunctions("startup")
+end
+
+-- [Bizhawk only] Executed each frame (60 frames per second)
+-- CAUTION: Avoid unnecessary calculations here, as this can easily affect performance.
+function CustomCode.inputCheckBizhawk()
+	CustomCode.execFunctions("inputCheckBizhawk")
+end
+
+-- [MGBA only] Executed each frame (60 frames per second)
+-- CAUTION: Avoid unnecessary calculations here, as this can easily affect performance.
+function CustomCode.inputCheckMGBA()
+	CustomCode.execFunctions("inputCheckMGBA")
+end
+
+-- Executed each frame, after most data from game memory is read in but before any natural redraw events occur
+-- CAUTION: Avoid code here if possible, as this can easily affect performance. Most Tracker updates occur at 30-frame intervals, some at 10-frame.
+function CustomCode.afterEachFrame()
+	CustomCode.execFunctions("afterEachFrame")
+end
+
+-- Executed once every 30 frames, after most data from game memory is read in
+function CustomCode.afterProgramDataUpdate()
+	CustomCode.execFunctions("afterProgramDataUpdate")
+end
+
+-- Executed once every 30 frames, after any battle-related data from game memory is read in
+function CustomCode.afterBattleDataUpdate()
+	CustomCode.execFunctions("afterBattleDataUpdate")
+end
+
+-- Executed once every 30 frames or after any redraw event is scheduled (e.g. most button presses)
+function CustomCode.afterRedraw()
+	CustomCode.execFunctions("afterRedraw")
+end
+
+-- Executed after a new battle begins (wild or trainer), and only once per battle
+function CustomCode.afterBattleBegins()
+	CustomCode.execFunctions("afterBattleBegins")
+end
+
+-- Executed after a battle ends, and only once per battle
+function CustomCode.afterBattleEnds()
+	CustomCode.execFunctions("afterBattleEnds")
+end
+
+-- Executed before a button is clicked, and only once per button
+-- button: the button object being clicked
+function CustomCode.beforeButtonClicked(button)
+	CustomCode.execFunctions("beforeButtonClicked", button)
+end
```

Comment:
I feel like the wording of this could potentially be slightly confusing, as it's worded as "before a button is clicked" however looking below at `Input.lua` I can see it's actually *after* the user has clicked on the button, but before the button's "onclick" function is carried out from clicking on it. Maybe something like "Executed when a button is clicked, before the button's functionality is carried out"

The wording "once per button" I think is the most confusing here as it implies it will only ever run once for a given button, when in actuality it runs once for that button *per click of the button* as I don't see anything in these changes that stops it running on successive clicks

Realistically it probably makes no difference though, only misinterpretation I can think of is someone somehow thinking it hooks into the button's initial creation and lets them see its initial state and not actually relating to running alongside its onclick functionality at all

Reply:
Good notes, i'll update the function name and/or description soon.